tests/unit/plugins/test_event_consumer.py

```diff
@@ -15,23 +15,549 @@
+        await consumer.cleanup(event_msg)
+
+    pubsub_msg.ack.assert_called_once_with()
+
+    exp_entry = {
+        'timestamp': '2018-01-01T11:30:00.000000Z',
+        'plugin': 'consume',
+        'message': 'Acknowledged message in Pub/Sub.',
+    }
+    assert exp_entry in event_msg.history_log
+    mock_event.set.assert_called_once_with()
+    mock_thread.stop.assert_called_once_with()
+    assert 2 == len(caplog.records)
+
+
+#####
+# _GPSThread tests
+#####
+def test_create_gpsthread(thread, event):
+    gps_thread = event_consumer._GPSThread(event)
+
+    assert not gps_thread.loop
+    assert not gps_thread.thread_id
+    assert event is gps_thread.event
+
+
+class CustomLoop(asyncio.BaseEventLoop):
+    def run_forever(self):
+        # so we don't _actually_ run forever
+        pass
+
+    def stop(self):
+        pass
+
+    def close(self):
+        pass
+
+    def call_soon_threadsafe(self, *args, **kwargs):
+        pass
+
+
+@pytest.fixture
+def custom_event_loop():
+    loop = CustomLoop()
+    yield loop
+    loop.close()
+
+
+def test_gpsthread_run(thread, event, custom_event_loop, monkeypatch):
+    monkeypatch.setattr(NEW_EV_PATCH, lambda: custom_event_loop)
+
+    gps_thread = event_consumer._GPSThread(event)
+    gps_thread.run()
+
+    assert custom_event_loop is gps_thread.loop
+    current_id = threading.current_thread()
+    assert current_id == gps_thread.thread_id
+
+
+def test_gpsthread_stop(thread, event, custom_event_loop):
+    gps_thread = event_consumer._GPSThread(event)
+    gps_thread.loop = custom_event_loop
+
+    gps_thread.stop()
+
+    assert not gps_thread.thread_id
+    assert not custom_event_loop.is_running()
+
+
+# different fixture than custom_event_loop because two loops are being
+# used, one for running other tests, and this one for the _GPSThread to
+# schedule tasks on
+@pytest.fixture
+def new_event_loop(event_loop):
+    policy = asyncio.get_event_loop_policy()
+    res = policy.new_event_loop()
+    asyncio.set_event_loop(res)
+    res._stop = res.stop
+    res.stop = lambda: None
+
+    yield res
+
+    res._stop()
+
+
+# NOTE: This test has some sort of race condition that on some runs
+#       (1 of every 5?) the loop will stop before the `noop` coro is
+#       awaited, so a warning will be emitted. This is okay since we're
+#       not testing that bit
+def test_gpsthread_add_task_outside_thread(thread, event, new_event_loop,
+                                           monkeypatch):
+    monkeypatch.setattr(NEW_EV_PATCH, lambda: new_event_loop)
+    gps_thread = event_consumer._GPSThread(event)
+    gps_thread.start()
+    event.wait()
+
+    async def noop():
+        return asyncio.Future()
+
+    future = gps_thread.add_task(noop())
+
+    assert isinstance(future, asyncio.Task)
+
+    # cleanup
+    # attempting to give time for the noop task to complete to avoid
+    # annoying & unactionable warnings
+    time.sleep(.3)
```

Comment:
Is there a way to signal that the noop task is done (maybe a threading.Event) to make this more deterministic?

Reply:
This actually doesn't do what I hoped. We still get warnings inconsistently either way (╯°□°）╯︵ ┻━┻ I'm just going to remove this bit.